// regexp/index.js
function parse(n) {
  if ("string" != typeof n) {
    var t = new TypeError("The regexp to parse must be represented as a string.");
    throw t;
  }
  index = 1, cgs = {}, result = parser.parse(n);
  for (var l = 0; l < n.length; l++) cgs[l] && (cgs[l].index = index++);
  return result;
}

function Token(n) {
  this.type = n, this.offset = Token.offset(), this.text = Token.text();
}

function Alternate(n, t) {
  Token.call(this, "alternate"), this.left = n, this.right = t;
}

function Match(n) {
  Token.call(this, "match"), this.body = n.filter(Boolean);
}

function Group(n, t) {
  Token.call(this, n), this.body = t;
}

function CaptureGroup(n, t) {
  Group.call(this, "capture-group"), cgs[this.offset] = this, this.body = n, t && (this.name = t[0] + t[1].join(""));
}

function Quantified(n, t) {
  Token.call(this, "quantified"), this.body = n, this.quantifier = t;
}

function Quantifier(n, t) {
  Token.call(this, "quantifier"), this.min = n, this.max = t, this.greedy = !0;
}

function CharSet(n, t) {
  Token.call(this, "charset"), this.invert = n, this.body = t;
}

function CharacterRange(n, t) {
  Token.call(this, "range"), this.start = n, this.end = t;
}

function CharacterClass(n, t) {
  Token.call(this, "charclass"), this.start = n, this.end = t;
}

function Literal(n) {
  Token.call(this, "literal"), this.body = n, this.escaped = this.body != this.text;
}

function Unicode(n) {
  Token.call(this, "unicode"), this.code = n.toUpperCase();
}

function UnicodeCategory(n, t) {
  Token.call(this, "unicode-category"), this.code = n, this.invert = t;
}

function Hex(n) {
  Token.call(this, "hex"), this.code = n.toUpperCase();
}

function Octal(n) {
  Token.call(this, "octal"), this.code = n.toUpperCase();
}

function BackReference(n) {
  Token.call(this, "back-reference"), this.code = n.toUpperCase();
}

function ControlCharacter(n) {
  Token.call(this, "control-character"), this.code = n.toUpperCase();
}

var parser = function() {
  function n(n, t) {
    function l() {
      this.constructor = n;
    }
    l.prototype = t.prototype, n.prototype = new l();
  }
  function t(n, t, l, r, u) {
    function e(n, t) {
      function l(n) {
        function t(n) {
          return n.charCodeAt(0).toString(16).toUpperCase();
        }
        return n.replace(/\\/g, "\\\\").replace(/"/g, '\\"').replace(/\x08/g, "\\b").replace(/\t/g, "\\t").replace(/\n/g, "\\n").replace(/\f/g, "\\f").replace(/\r/g, "\\r").replace(/[\x00-\x07\x0B\x0E\x0F]/g, function(n) {
          return "\\x0" + t(n);
        }).replace(/[\x10-\x1F\x80-\xFF]/g, function(n) {
          return "\\x" + t(n);
        }).replace(/[\u0180-\u0FFF]/g, function(n) {
          return "\\u0" + t(n);
        }).replace(/[\u1080-\uFFFF]/g, function(n) {
          return "\\u" + t(n);
        });
      }
      var r, u;
      switch (n.length) {
       case 0:
        r = "end of input";
        break;

       case 1:
        r = n[0];
        break;

       default:
        r = n.slice(0, -1).join(", ") + " or " + n[n.length - 1];
      }
      return u = t ? '"' + l(t) + '"' : "end of input", "Expected " + r + " but " + u + " found.";
    }
    this.expected = n, this.found = t, this.offset = l, this.line = r, this.column = u, 
    this.name = "SyntaxError", this.message = e(n, t);
  }
  function l(n) {
    function l() {
      return n.substring(Ce, de);
    }
    function r() {
      return Ce;
    }
    function u(t) {
      function l(t, l, r) {
        var u, e;
        for (u = l; r > u; u++) e = n.charAt(u), "\n" === e ? (t.seenCR || t.line++, t.column = 1, 
        t.seenCR = !1) : "\r" === e || "\u2028" === e || "\u2029" === e ? (t.line++, t.column = 1, 
        t.seenCR = !0) : (t.column++, t.seenCR = !1);
      }
      return Ae !== t && (Ae > t && (Ae = 0, be = {
        line: 1,
        column: 1,
        seenCR: !1
      }), l(be, Ae, t), Ae = t), be;
    }
    function e(n) {
      ve > de || (de > ve && (ve = de, ke = []), ke.push(n));
    }
    function o(n) {
      var t = 0;
      for (n.sort(); t < n.length; ) n[t - 1] === n[t] ? n.splice(t, 1) : t++;
    }
    function c() {
      var t, l, r, u, o;
      return t = de, l = a(), null !== l ? (r = de, 124 === n.charCodeAt(de) ? (u = xt, 
      de++) : (u = null, 0 === ye && e(Tt)), null !== u ? (o = c(), null !== o ? (u = [ u, o ], 
      r = u) : (de = r, r = gt)) : (de = r, r = gt), null === r && (r = wt), null !== r ? (Ce = t, 
      l = Gt(l, r), null === l ? (de = t, t = l) : t = l) : (de = t, t = gt)) : (de = t, 
      t = gt), t;
    }
    function a() {
      var n, t, l, r, u;
      if (n = de, t = s(), null === t && (t = wt), null !== t) if (l = de, ye++, r = h(), 
      ye--, null === r ? l = wt : (de = l, l = gt), null !== l) {
        for (r = [], u = G(), null === u && (u = p(), null === u && (u = i())); null !== u; ) r.push(u), 
        u = G(), null === u && (u = p(), null === u && (u = i()));
        null !== r ? (u = f(), null === u && (u = wt), null !== u ? (Ce = n, t = mt(t, r, u), 
        null === t ? (de = n, n = t) : n = t) : (de = n, n = gt)) : (de = n, n = gt);
      } else de = n, n = gt; else de = n, n = gt;
      return n;
    }
    function i() {
      var n;
      return n = x(), null === n && (n = B(), null === n && (n = H())), n;
    }
    function s() {
      var t, l;
      return t = de, 94 === n.charCodeAt(de) ? (l = Rt, de++) : (l = null, 0 === ye && e(jt)), 
      null !== l && (Ce = t, l = Ot()), null === l ? (de = t, t = l) : t = l, t;
    }
    function f() {
      var t, l;
      return t = de, 36 === n.charCodeAt(de) ? (l = Ut, de++) : (l = null, 0 === ye && e(Qt)), 
      null !== l && (Ce = t, l = Ft()), null === l ? (de = t, t = l) : t = l, t;
    }
    function p() {
      var n, t, l;
      return n = de, t = i(), null !== t ? (l = h(), null !== l ? (Ce = n, t = St(t, l), 
      null === t ? (de = n, n = t) : n = t) : (de = n, n = gt)) : (de = n, n = gt), n;
    }
    function h() {
      var n, t, l;
      return ye++, n = de, t = d(), null !== t ? (l = g(), null === l && (l = wt), null !== l ? (Ce = n, 
      t = Mt(t, l), null === t ? (de = n, n = t) : n = t) : (de = n, n = gt)) : (de = n, 
      n = gt), ye--, null === n && (t = null, 0 === ye && e(Bt)), n;
    }
    function d() {
      var n;
      return n = C(), null === n && (n = A(), null === n && (n = b(), null === n && (n = v(), 
      null === n && (n = k(), null === n && (n = y()))))), n;
    }
    function C() {
      var t, l, r, u, o, c;
      return t = de, 123 === n.charCodeAt(de) ? (l = zt, de++) : (l = null, 0 === ye && e(Lt)), 
      null !== l ? (r = w(), null !== r ? (44 === n.charCodeAt(de) ? (u = Zt, de++) : (u = null, 
      0 === ye && e(Et)), null !== u ? (o = w(), null !== o ? (125 === n.charCodeAt(de) ? (c = Ht, 
      de++) : (c = null, 0 === ye && e(_t)), null !== c ? (Ce = t, l = Pt(r, o), null === l ? (de = t, 
      t = l) : t = l) : (de = t, t = gt)) : (de = t, t = gt)) : (de = t, t = gt)) : (de = t, 
      t = gt)) : (de = t, t = gt), t;
    }
    function A() {
      var t, l, r, u;
      return t = de, 123 === n.charCodeAt(de) ? (l = zt, de++) : (l = null, 0 === ye && e(Lt)), 
      null !== l ? (r = w(), null !== r ? (n.substr(de, 2) === $t ? (u = $t, de += 2) : (u = null, 
      0 === ye && e(qt)), null !== u ? (Ce = t, l = Dt(r), null === l ? (de = t, t = l) : t = l) : (de = t, 
      t = gt)) : (de = t, t = gt)) : (de = t, t = gt), t;
    }
    function b() {
      var t, l, r, u;
      return t = de, 123 === n.charCodeAt(de) ? (l = zt, de++) : (l = null, 0 === ye && e(Lt)), 
      null !== l ? (r = w(), null !== r ? (125 === n.charCodeAt(de) ? (u = Ht, de++) : (u = null, 
      0 === ye && e(_t)), null !== u ? (Ce = t, l = Wt(r), null === l ? (de = t, t = l) : t = l) : (de = t, 
      t = gt)) : (de = t, t = gt)) : (de = t, t = gt), t;
    }
    function v() {
      var t, l;
      return t = de, 43 === n.charCodeAt(de) ? (l = It, de++) : (l = null, 0 === ye && e(Jt)), 
      null !== l && (Ce = t, l = Kt()), null === l ? (de = t, t = l) : t = l, t;
    }
    function k() {
      var t, l;
      return t = de, 42 === n.charCodeAt(de) ? (l = Nt, de++) : (l = null, 0 === ye && e(Vt)), 
      null !== l && (Ce = t, l = Xt()), null === l ? (de = t, t = l) : t = l, t;
    }
    function y() {
      var t, l;
      return t = de, 63 === n.charCodeAt(de) ? (l = Yt, de++) : (l = null, 0 === ye && e(nl)), 
      null !== l && (Ce = t, l = tl()), null === l ? (de = t, t = l) : t = l, t;
    }
    function g() {
      var t;
      return 63 === n.charCodeAt(de) ? (t = Yt, de++) : (t = null, 0 === ye && e(nl)), 
      t;
    }
    function w() {
      var t, l, r;
      if (t = de, l = [], ll.test(n.charAt(de)) ? (r = n.charAt(de), de++) : (r = null, 
      0 === ye && e(rl)), null !== r) for (;null !== r; ) l.push(r), ll.test(n.charAt(de)) ? (r = n.charAt(de), 
      de++) : (r = null, 0 === ye && e(rl)); else l = gt;
      return null !== l && (Ce = t, l = ul(l)), null === l ? (de = t, t = l) : t = l, 
      t;
    }
    function x() {
      var t, l, r, u;
      return t = de, 40 === n.charCodeAt(de) ? (l = el, de++) : (l = null, 0 === ye && e(ol)), 
      null !== l ? (r = j(), null === r && (r = O(), null === r && (r = U(), null === r && (r = Q(), 
      null === r && (r = F(), null === r && (r = T(), null === r && (r = G(), null === r && (r = R(), 
      null === r && (r = m())))))))), null !== r ? (41 === n.charCodeAt(de) ? (u = cl, 
      de++) : (u = null, 0 === ye && e(al)), null !== u ? (Ce = t, l = il(r), null === l ? (de = t, 
      t = l) : t = l) : (de = t, t = gt)) : (de = t, t = gt)) : (de = t, t = gt), t;
    }
    function T() {
      var t, l, r, u, o;
      if (t = de, 63 === n.charCodeAt(de) ? (l = Yt, de++) : (l = null, 0 === ye && e(nl)), 
      null !== l) {
        if (r = [], sl.test(n.charAt(de)) ? (u = n.charAt(de), de++) : (u = null, 0 === ye && e(fl)), 
        null !== u) for (;null !== u; ) r.push(u), sl.test(n.charAt(de)) ? (u = n.charAt(de), 
        de++) : (u = null, 0 === ye && e(fl)); else r = gt;
        null !== r ? (58 === n.charCodeAt(de) ? (u = pl, de++) : (u = null, 0 === ye && e(hl)), 
        null !== u ? (o = c(), null !== o ? (Ce = t, l = dl(r, o), null === l ? (de = t, 
        t = l) : t = l) : (de = t, t = gt)) : (de = t, t = gt)) : (de = t, t = gt);
      } else de = t, t = gt;
      return t;
    }
    function G() {
      var t, l, r, u;
      if (t = de, 63 === n.charCodeAt(de) ? (l = Yt, de++) : (l = null, 0 === ye && e(nl)), 
      null !== l) {
        if (r = [], sl.test(n.charAt(de)) ? (u = n.charAt(de), de++) : (u = null, 0 === ye && e(fl)), 
        null !== u) for (;null !== u; ) r.push(u), sl.test(n.charAt(de)) ? (u = n.charAt(de), 
        de++) : (u = null, 0 === ye && e(fl)); else r = gt;
        null !== r ? (Ce = t, l = Cl(r), null === l ? (de = t, t = l) : t = l) : (de = t, 
        t = gt);
      } else de = t, t = gt;
      return t;
    }
    function m() {
      var n, t;
      return n = de, t = c(), null !== t && (Ce = n, t = Al(t)), null === t ? (de = n, 
      n = t) : n = t, n;
    }
    function R() {
      var t, l, r;
      return t = de, n.substr(de, 2) === bl ? (l = bl, de += 2) : (l = null, 0 === ye && e(vl)), 
      null !== l ? (r = c(), null !== r ? (Ce = t, l = kl(r), null === l ? (de = t, t = l) : t = l) : (de = t, 
      t = gt)) : (de = t, t = gt), t;
    }
    function j() {
      var t, l, r;
      return t = de, n.substr(de, 2) === yl ? (l = yl, de += 2) : (l = null, 0 === ye && e(gl)), 
      null !== l ? (r = c(), null !== r ? (Ce = t, l = wl(r), null === l ? (de = t, t = l) : t = l) : (de = t, 
      t = gt)) : (de = t, t = gt), t;
    }
    function O() {
      var t, l, r;
      return t = de, n.substr(de, 2) === xl ? (l = xl, de += 2) : (l = null, 0 === ye && e(Tl)), 
      null !== l ? (r = c(), null !== r ? (Ce = t, l = Gl(r), null === l ? (de = t, t = l) : t = l) : (de = t, 
      t = gt)) : (de = t, t = gt), t;
    }
    function U() {
      var t, l, r;
      return t = de, n.substr(de, 3) === ml ? (l = ml, de += 3) : (l = null, 0 === ye && e(Rl)), 
      null !== l ? (r = c(), null !== r ? (Ce = t, l = jl(r), null === l ? (de = t, t = l) : t = l) : (de = t, 
      t = gt)) : (de = t, t = gt), t;
    }
    function Q() {
      var t, l, r;
      return t = de, n.substr(de, 3) === Ol ? (l = Ol, de += 3) : (l = null, 0 === ye && e(Ul)), 
      null !== l ? (r = c(), null !== r ? (Ce = t, l = Ql(r), null === l ? (de = t, t = l) : t = l) : (de = t, 
      t = gt)) : (de = t, t = gt), t;
    }
    function F() {
      var t, l, r, u, o, a;
      if (t = de, l = S(), null !== l) {
        if (r = de, Fl.test(n.charAt(de)) ? (u = n.charAt(de), de++) : (u = null, 0 === ye && e(Sl)), 
        null !== u) {
          for (o = [], Bl.test(n.charAt(de)) ? (a = n.charAt(de), de++) : (a = null, 0 === ye && e(Ml)); null !== a; ) o.push(a), 
          Bl.test(n.charAt(de)) ? (a = n.charAt(de), de++) : (a = null, 0 === ye && e(Ml));
          null !== o ? (u = [ u, o ], r = u) : (de = r, r = gt);
        } else de = r, r = gt;
        null !== r ? (62 === n.charCodeAt(de) ? (u = zl, de++) : (u = null, 0 === ye && e(Ll)), 
        null !== u ? (o = c(), null !== o ? (Ce = t, l = Zl(r, o), null === l ? (de = t, 
        t = l) : t = l) : (de = t, t = gt)) : (de = t, t = gt)) : (de = t, t = gt);
      } else de = t, t = gt;
      return t;
    }
    function S() {
      var t;
      return n.substr(de, 3) === El ? (t = El, de += 3) : (t = null, 0 === ye && e(Hl)), 
      null === t && (n.substr(de, 2) === _l ? (t = _l, de += 2) : (t = null, 0 === ye && e(Pl))), 
      t;
    }
    function B() {
      var t, l, r, u, o;
      if (ye++, t = de, 91 === n.charCodeAt(de) ? (l = ql, de++) : (l = null, 0 === ye && e(Dl)), 
      null !== l) if (94 === n.charCodeAt(de) ? (r = Rt, de++) : (r = null, 0 === ye && e(jt)), 
      null === r && (r = wt), null !== r) {
        for (u = [], o = z(), null === o && (o = M(), null === o && (o = L())); null !== o; ) u.push(o), 
        o = z(), null === o && (o = M(), null === o && (o = L()));
        null !== u ? (93 === n.charCodeAt(de) ? (o = Wl, de++) : (o = null, 0 === ye && e(Il)), 
        null !== o ? (Ce = t, l = Jl(r, u), null === l ? (de = t, t = l) : t = l) : (de = t, 
        t = gt)) : (de = t, t = gt);
      } else de = t, t = gt; else de = t, t = gt;
      return ye--, null === t && (l = null, 0 === ye && e($l)), t;
    }
    function M() {
      var t, l, r, u;
      return ye++, t = de, l = L(), null !== l ? (45 === n.charCodeAt(de) ? (r = Nl, de++) : (r = null, 
      0 === ye && e(Vl)), null !== r ? (u = L(), null !== u ? (Ce = t, l = Xl(l, u), null === l ? (de = t, 
      t = l) : t = l) : (de = t, t = gt)) : (de = t, t = gt)) : (de = t, t = gt), ye--, 
      null === t && (l = null, 0 === ye && e(Kl)), t;
    }
    function z() {
      var t, l, r, u;
      return ye++, t = de, n.substr(de, 2) === nr ? (l = nr, de += 2) : (l = null, 0 === ye && e(tr)), 
      null !== l ? (n.substr(de, 5) === lr ? (r = lr, de += 5) : (r = null, 0 === ye && e(rr)), 
      null === r && (n.substr(de, 5) === ur ? (r = ur, de += 5) : (r = null, 0 === ye && e(er)), 
      null === r && (n.substr(de, 5) === or ? (r = or, de += 5) : (r = null, 0 === ye && e(cr)), 
      null === r && (n.substr(de, 5) === ar ? (r = ar, de += 5) : (r = null, 0 === ye && e(ir)), 
      null === r && (n.substr(de, 5) === sr ? (r = sr, de += 5) : (r = null, 0 === ye && e(fr)), 
      null === r && (n.substr(de, 5) === pr ? (r = pr, de += 5) : (r = null, 0 === ye && e(hr)), 
      null === r && (n.substr(de, 5) === dr ? (r = dr, de += 5) : (r = null, 0 === ye && e(Cr)), 
      null === r && (n.substr(de, 5) === Ar ? (r = Ar, de += 5) : (r = null, 0 === ye && e(br)), 
      null === r && (n.substr(de, 5) === vr ? (r = vr, de += 5) : (r = null, 0 === ye && e(kr)), 
      null === r && (n.substr(de, 5) === yr ? (r = yr, de += 5) : (r = null, 0 === ye && e(gr)), 
      null === r && (n.substr(de, 5) === wr ? (r = wr, de += 5) : (r = null, 0 === ye && e(xr)), 
      null === r && (n.substr(de, 6) === Tr ? (r = Tr, de += 6) : (r = null, 0 === ye && e(Gr))))))))))))), 
      null !== r ? (n.substr(de, 2) === mr ? (u = mr, de += 2) : (u = null, 0 === ye && e(Rr)), 
      null !== u ? (Ce = t, l = jr(r), null === l ? (de = t, t = l) : t = l) : (de = t, 
      t = gt)) : (de = t, t = gt)) : (de = t, t = gt), ye--, null === t && (l = null, 
      0 === ye && e(Yl)), t;
    }
    function L() {
      var n, t;
      return ye++, n = E(), null === n && (n = Z()), ye--, null === n && (t = null, 0 === ye && e(Or)), 
      n;
    }
    function Z() {
      var t, l;
      return t = de, Ur.test(n.charAt(de)) ? (l = n.charAt(de), de++) : (l = null, 0 === ye && e(Qr)), 
      null !== l && (Ce = t, l = Fr(l)), null === l ? (de = t, t = l) : t = l, t;
    }
    function E() {
      var n;
      return n = q(), null === n && (n = at(), null === n && (n = V(), null === n && (n = X(), 
      null === n && (n = Y(), null === n && (n = nt(), null === n && (n = tt(), null === n && (n = lt(), 
      null === n && (n = rt(), null === n && (n = ut(), null === n && (n = et(), null === n && (n = ot(), 
      null === n && (n = ct(), null === n && (n = st(), null === n && (n = ft(), null === n && (n = pt(), 
      null === n && (n = Ct(), null === n && (n = At()))))))))))))))))), n;
    }
    function H() {
      var n;
      return n = _(), null === n && (n = $(), null === n && (n = P())), n;
    }
    function _() {
      var t, l;
      return t = de, 46 === n.charCodeAt(de) ? (l = Sr, de++) : (l = null, 0 === ye && e(Br)), 
      null !== l && (Ce = t, l = Mr()), null === l ? (de = t, t = l) : t = l, t;
    }
    function P() {
      var t, l;
      return ye++, t = de, Lr.test(n.charAt(de)) ? (l = n.charAt(de), de++) : (l = null, 
      0 === ye && e(Zr)), null !== l && (Ce = t, l = Fr(l)), null === l ? (de = t, t = l) : t = l, 
      ye--, null === t && (l = null, 0 === ye && e(zr)), t;
    }
    function $() {
      var n;
      return n = D(), null === n && (n = N(), null === n && (n = W(), null === n && (n = I(), 
      null === n && (n = J(), null === n && (n = K(), null === n && (n = at(), null === n && (n = V(), 
      null === n && (n = X(), null === n && (n = Y(), null === n && (n = nt(), null === n && (n = tt(), 
      null === n && (n = lt(), null === n && (n = rt(), null === n && (n = ut(), null === n && (n = et(), 
      null === n && (n = ot(), null === n && (n = ct(), null === n && (n = it(), null === n && (n = st(), 
      null === n && (n = ft(), null === n && (n = pt(), null === n && (n = ht(), null === n && (n = dt(), 
      null === n && (n = Ct(), null === n && (n = At()))))))))))))))))))))))))), n;
    }
    function q() {
      var t, l;
      return t = de, n.substr(de, 2) === Er ? (l = Er, de += 2) : (l = null, 0 === ye && e(Hr)), 
      null !== l && (Ce = t, l = _r()), null === l ? (de = t, t = l) : t = l, t;
    }
    function D() {
      var t, l;
      return t = de, n.substr(de, 2) === Er ? (l = Er, de += 2) : (l = null, 0 === ye && e(Hr)), 
      null !== l && (Ce = t, l = Pr()), null === l ? (de = t, t = l) : t = l, t;
    }
    function W() {
      var t, l;
      return t = de, n.substr(de, 2) === $r ? (l = $r, de += 2) : (l = null, 0 === ye && e(qr)), 
      null !== l && (Ce = t, l = Dr()), null === l ? (de = t, t = l) : t = l, t;
    }
    function I() {
      var t, l;
      return t = de, n.substr(de, 2) === Wr ? (l = Wr, de += 2) : (l = null, 0 === ye && e(Ir)), 
      null !== l && (Ce = t, l = Jr()), null === l ? (de = t, t = l) : t = l, t;
    }
    function J() {
      var t, l;
      return t = de, n.substr(de, 2) === Kr ? (l = Kr, de += 2) : (l = null, 0 === ye && e(Nr)), 
      null !== l && (Ce = t, l = Vr()), null === l ? (de = t, t = l) : t = l, t;
    }
    function K() {
      var t, l;
      return t = de, n.substr(de, 2) === Xr ? (l = Xr, de += 2) : (l = null, 0 === ye && e(Yr)), 
      null !== l && (Ce = t, l = nu()), null === l ? (de = t, t = l) : t = l, t;
    }
    function N() {
      var t, l;
      return t = de, n.substr(de, 2) === tu ? (l = tu, de += 2) : (l = null, 0 === ye && e(lu)), 
      null !== l && (Ce = t, l = ru()), null === l ? (de = t, t = l) : t = l, t;
    }
    function V() {
      var t, l;
      return t = de, n.substr(de, 2) === uu ? (l = uu, de += 2) : (l = null, 0 === ye && e(eu)), 
      null !== l && (Ce = t, l = ou()), null === l ? (de = t, t = l) : t = l, t;
    }
    function X() {
      var t, l;
      return t = de, n.substr(de, 2) === cu ? (l = cu, de += 2) : (l = null, 0 === ye && e(au)), 
      null !== l && (Ce = t, l = iu()), null === l ? (de = t, t = l) : t = l, t;
    }
    function Y() {
      var t, l;
      return t = de, n.substr(de, 2) === su ? (l = su, de += 2) : (l = null, 0 === ye && e(fu)), 
      null !== l && (Ce = t, l = pu()), null === l ? (de = t, t = l) : t = l, t;
    }
    function nt() {
      var t, l;
      return t = de, n.substr(de, 2) === hu ? (l = hu, de += 2) : (l = null, 0 === ye && e(du)), 
      null !== l && (Ce = t, l = Cu()), null === l ? (de = t, t = l) : t = l, t;
    }
    function tt() {
      var t, l;
      return t = de, n.substr(de, 2) === Au ? (l = Au, de += 2) : (l = null, 0 === ye && e(bu)), 
      null !== l && (Ce = t, l = vu()), null === l ? (de = t, t = l) : t = l, t;
    }
    function lt() {
      var t, l;
      return t = de, n.substr(de, 2) === ku ? (l = ku, de += 2) : (l = null, 0 === ye && e(yu)), 
      null !== l && (Ce = t, l = gu()), null === l ? (de = t, t = l) : t = l, t;
    }
    function rt() {
      var t, l;
      return t = de, n.substr(de, 2) === wu ? (l = wu, de += 2) : (l = null, 0 === ye && e(xu)), 
      null !== l && (Ce = t, l = Tu()), null === l ? (de = t, t = l) : t = l, t;
    }
    function ut() {
      var t, l;
      return t = de, n.substr(de, 2) === Gu ? (l = Gu, de += 2) : (l = null, 0 === ye && e(mu)), 
      null !== l && (Ce = t, l = Ru()), null === l ? (de = t, t = l) : t = l, t;
    }
    function et() {
      var t, l;
      return t = de, n.substr(de, 2) === ju ? (l = ju, de += 2) : (l = null, 0 === ye && e(Ou)), 
      null !== l && (Ce = t, l = Uu()), null === l ? (de = t, t = l) : t = l, t;
    }
    function ot() {
      var t, l;
      return t = de, n.substr(de, 2) === Qu ? (l = Qu, de += 2) : (l = null, 0 === ye && e(Fu)), 
      null !== l && (Ce = t, l = Su()), null === l ? (de = t, t = l) : t = l, t;
    }
    function ct() {
      var t, l;
      return t = de, n.substr(de, 2) === Bu ? (l = Bu, de += 2) : (l = null, 0 === ye && e(Mu)), 
      null !== l && (Ce = t, l = zu()), null === l ? (de = t, t = l) : t = l, t;
    }
    function at() {
      var t, l, r;
      return t = de, n.substr(de, 2) === Lu ? (l = Lu, de += 2) : (l = null, 0 === ye && e(Zu)), 
      null !== l ? (n.length > de ? (r = n.charAt(de), de++) : (r = null, 0 === ye && e(Eu)), 
      null !== r ? (Ce = t, l = Hu(r), null === l ? (de = t, t = l) : t = l) : (de = t, 
      t = gt)) : (de = t, t = gt), t;
    }
    function it() {
      var t, l, r;
      return t = de, 92 === n.charCodeAt(de) ? (l = _u, de++) : (l = null, 0 === ye && e(Pu)), 
      null !== l ? ($u.test(n.charAt(de)) ? (r = n.charAt(de), de++) : (r = null, 0 === ye && e(qu)), 
      null !== r ? (Ce = t, l = Du(r), null === l ? (de = t, t = l) : t = l) : (de = t, 
      t = gt)) : (de = t, t = gt), t;
    }
    function st() {
      var t, l, r, u;
      if (t = de, n.substr(de, 2) === Wu ? (l = Wu, de += 2) : (l = null, 0 === ye && e(Iu)), 
      null !== l) {
        if (r = [], Ju.test(n.charAt(de)) ? (u = n.charAt(de), de++) : (u = null, 0 === ye && e(Ku)), 
        null !== u) for (;null !== u; ) r.push(u), Ju.test(n.charAt(de)) ? (u = n.charAt(de), 
        de++) : (u = null, 0 === ye && e(Ku)); else r = gt;
        null !== r ? (Ce = t, l = Nu(r), null === l ? (de = t, t = l) : t = l) : (de = t, 
        t = gt);
      } else de = t, t = gt;
      return t;
    }
    function ft() {
      var t, l, r, u;
      if (t = de, n.substr(de, 2) === Vu ? (l = Vu, de += 2) : (l = null, 0 === ye && e(Xu)), 
      null !== l) {
        if (r = [], Yu.test(n.charAt(de)) ? (u = n.charAt(de), de++) : (u = null, 0 === ye && e(ne)), 
        null !== u) for (;null !== u; ) r.push(u), Yu.test(n.charAt(de)) ? (u = n.charAt(de), 
        de++) : (u = null, 0 === ye && e(ne)); else r = gt;
        null !== r ? (Ce = t, l = te(r), null === l ? (de = t, t = l) : t = l) : (de = t, 
        t = gt);
      } else de = t, t = gt;
      return t;
    }
    function pt() {
      var t, l, r, u;
      if (t = de, n.substr(de, 2) === le ? (l = le, de += 2) : (l = null, 0 === ye && e(re)), 
      null !== l) {
        if (r = [], Yu.test(n.charAt(de)) ? (u = n.charAt(de), de++) : (u = null, 0 === ye && e(ne)), 
        null !== u) for (;null !== u; ) r.push(u), Yu.test(n.charAt(de)) ? (u = n.charAt(de), 
        de++) : (u = null, 0 === ye && e(ne)); else r = gt;
        null !== r ? (Ce = t, l = ue(r), null === l ? (de = t, t = l) : t = l) : (de = t, 
        t = gt);
      } else de = t, t = gt;
      return t;
    }
    function ht() {
      var t, l, r, u;
      if (t = de, n.substr(de, 3) === ee ? (l = ee, de += 3) : (l = null, 0 === ye && e(oe)), 
      null !== l) {
        if (r = [], ce.test(n.charAt(de)) ? (u = n.charAt(de), de++) : (u = null, 0 === ye && e(ae)), 
        null !== u) for (;null !== u; ) r.push(u), ce.test(n.charAt(de)) ? (u = n.charAt(de), 
        de++) : (u = null, 0 === ye && e(ae)); else r = gt;
        null !== r ? (125 === n.charCodeAt(de) ? (u = Ht, de++) : (u = null, 0 === ye && e(_t)), 
        null !== u ? (Ce = t, l = ie(r), null === l ? (de = t, t = l) : t = l) : (de = t, 
        t = gt)) : (de = t, t = gt);
      } else de = t, t = gt;
      return t;
    }
    function dt() {
      var t, l, r, u;
      if (t = de, n.substr(de, 3) === se ? (l = se, de += 3) : (l = null, 0 === ye && e(fe)), 
      null !== l) {
        if (r = [], ce.test(n.charAt(de)) ? (u = n.charAt(de), de++) : (u = null, 0 === ye && e(ae)), 
        null !== u) for (;null !== u; ) r.push(u), ce.test(n.charAt(de)) ? (u = n.charAt(de), 
        de++) : (u = null, 0 === ye && e(ae)); else r = gt;
        null !== r ? (125 === n.charCodeAt(de) ? (u = Ht, de++) : (u = null, 0 === ye && e(_t)), 
        null !== u ? (Ce = t, l = pe(r), null === l ? (de = t, t = l) : t = l) : (de = t, 
        t = gt)) : (de = t, t = gt);
      } else de = t, t = gt;
      return t;
    }
    function Ct() {
      var t, l;
      return t = de, n.substr(de, 2) === Wu ? (l = Wu, de += 2) : (l = null, 0 === ye && e(Iu)), 
      null !== l && (Ce = t, l = he()), null === l ? (de = t, t = l) : t = l, t;
    }
    function At() {
      var t, l, r;
      return t = de, 92 === n.charCodeAt(de) ? (l = _u, de++) : (l = null, 0 === ye && e(Pu)), 
      null !== l ? (n.length > de ? (r = n.charAt(de), de++) : (r = null, 0 === ye && e(Eu)), 
      null !== r ? (Ce = t, l = Fr(r), null === l ? (de = t, t = l) : t = l) : (de = t, 
      t = gt)) : (de = t, t = gt), t;
    }
    var bt, vt = arguments.length > 1 ? arguments[1] : {}, kt = {
      regexp: c
    }, yt = c, gt = null, wt = "", xt = "|", Tt = '"|"', Gt = function(n, t) {
      return t ? new Alternate(n, t[1]) : n;
    }, mt = function(n, t, l) {
      return new Match([ n ].concat(t).concat([ l ]));
    }, Rt = "^", jt = '"^"', Ot = function() {
      return new Token("start");
    }, Ut = "$", Qt = '"$"', Ft = function() {
      return new Token("end");
    }, St = function(n, t) {
      return new Quantified(n, t);
    }, Bt = "Quantifier", Mt = function(n, t) {
      return t && (n.greedy = !1), n;
    }, zt = "{", Lt = '"{"', Zt = ",", Et = '","', Ht = "}", _t = '"}"', Pt = function(n, t) {
      return new Quantifier(n, t);
    }, $t = ",}", qt = '",}"', Dt = function(n) {
      return new Quantifier(n, 1/0);
    }, Wt = function(n) {
      return new Quantifier(n, n);
    }, It = "+", Jt = '"+"', Kt = function() {
      return new Quantifier(1, 1/0);
    }, Nt = "*", Vt = '"*"', Xt = function() {
      return new Quantifier(0, 1/0);
    }, Yt = "?", nl = '"?"', tl = function() {
      return new Quantifier(0, 1);
    }, ll = /^[0-9]/, rl = "[0-9]", ul = function(n) {
      return +n.join("");
    }, el = "(", ol = '"("', cl = ")", al = '")"', il = function(n) {
      return n;
    }, sl = /^[adluimsx\-]/, fl = "[adluimsx\\-]", pl = ":", hl = '":"', dl = function(n, t) {
      return new Group("non-capture-group", new Match([ new Group("flags", n), t ]));
    }, Cl = function(n) {
      return new Group("flags", n);
    }, Al = function(n) {
      return new CaptureGroup(n);
    }, bl = "?:", vl = '"?:"', kl = function(n) {
      return new Group("non-capture-group", n);
    }, yl = "?=", gl = '"?="', wl = function(n) {
      return new Group("positive-lookahead", n);
    }, xl = "?!", Tl = '"?!"', Gl = function(n) {
      return new Group("negative-lookahead", n);
    }, ml = "?<=", Rl = '"?<="', jl = function(n) {
      return new Group("positive-lookbehind", n);
    }, Ol = "?<!", Ul = '"?<!"', Ql = function(n) {
      return new Group("negative-lookbehind", n);
    }, Fl = /^[A-Za-z_]/, Sl = "[A-Za-z_]", Bl = /^[A-Za-z0-9_]/, Ml = "[A-Za-z0-9_]", zl = ">", Ll = '">"', Zl = function(n, t) {
      return new CaptureGroup(t, n);
    }, El = "?P<", Hl = '"?P<"', _l = "?<", Pl = '"?<"', $l = "CharacterSet", ql = "[", Dl = '"["', Wl = "]", Il = '"]"', Jl = function(n, t) {
      return new CharSet(!!n, t);
    }, Kl = "CharacterRange", Nl = "-", Vl = '"-"', Xl = function(n, t) {
      return new CharacterRange(n, t);
    }, Yl = "CharacterClass", nr = "[:", tr = '"[:"', lr = "alnum", rr = '"alnum"', ur = "alpha", er = '"alpha"', or = "blank", cr = '"blank"', ar = "cntrl", ir = '"cntrl"', sr = "digit", fr = '"digit"', pr = "lower", hr = '"lower"', dr = "upper", Cr = '"upper"', Ar = "graph", br = '"graph"', vr = "print", kr = '"print"', yr = "punct", gr = '"punct"', wr = "space", xr = '"space"', Tr = "xdigit", Gr = '"xdigit"', mr = ":]", Rr = '":]"', jr = function(n) {
      return new CharacterClass(n);
    }, Or = "Character", Ur = /^[^\\\]]/, Qr = "[^\\\\\\]]", Fr = function(n) {
      return new Literal(n);
    }, Sr = ".", Br = '"."', Mr = function() {
      return new Token("any-character");
    }, zr = "Literal", Lr = /^[^|\\.[()?+*$\^]/, Zr = "[^|\\\\.[()?+*$\\^]", Er = "\\b", Hr = '"\\\\b"', _r = function() {
      return new Token("backspace");
    }, Pr = function() {
      return new Token("word-boundary");
    }, $r = "\\A", qr = '"\\\\A"', Dr = function() {
      return new Token("begin-of-string");
    }, Wr = "\\Z", Ir = '"\\\\Z"', Jr = function() {
      return new Token("end-of-string-before-nl");
    }, Kr = "\\z", Nr = '"\\\\z"', Vr = function() {
      return new Token("end-of-string");
    }, Xr = "\\G", Yr = '"\\\\G"', nu = function() {
      return new Token("match-start-position");
    }, tu = "\\B", lu = '"\\\\B"', ru = function() {
      return new Token("non-word-boundary");
    }, uu = "\\d", eu = '"\\\\d"', ou = function() {
      return new Token("digit");
    }, cu = "\\D", au = '"\\\\D"', iu = function() {
      return new Token("non-digit");
    }, su = "\\f", fu = '"\\\\f"', pu = function() {
      return new Token("form-feed");
    }, hu = "\\n", du = '"\\\\n"', Cu = function() {
      return new Token("line-feed");
    }, Au = "\\r", bu = '"\\\\r"', vu = function() {
      return new Token("carriage-return");
    }, ku = "\\s", yu = '"\\\\s"', gu = function() {
      return new Token("white-space");
    }, wu = "\\S", xu = '"\\\\S"', Tu = function() {
      return new Token("non-white-space");
    }, Gu = "\\t", mu = '"\\\\t"', Ru = function() {
      return new Token("tab");
    }, ju = "\\v", Ou = '"\\\\v"', Uu = function() {
      return new Token("vertical-tab");
    }, Qu = "\\w", Fu = '"\\\\w"', Su = function() {
      return new Token("word");
    }, Bu = "\\W", Mu = '"\\\\W"', zu = function() {
      return new Token("non-word");
    }, Lu = "\\c", Zu = '"\\\\c"', Eu = "any character", Hu = function(n) {
      return new ControlCharacter(n);
    }, _u = "\\", Pu = '"\\\\"', $u = /^[1-9]/, qu = "[1-9]", Du = function(n) {
      return new BackReference(n);
    }, Wu = "\\0", Iu = '"\\\\0"', Ju = /^[0-7]/, Ku = "[0-7]", Nu = function(n) {
      return new Octal(n.join(""));
    }, Vu = "\\x", Xu = '"\\\\x"', Yu = /^[0-9a-fA-F]/, ne = "[0-9a-fA-F]", te = function(n) {
      return new Hex(n.join(""));
    }, le = "\\u", re = '"\\\\u"', ue = function(n) {
      return new Unicode(n.join(""));
    }, ee = "\\p{", oe = '"\\\\p{"', ce = /^[0-9a-zA-Z_]/, ae = "[0-9a-zA-Z_]", ie = function(n) {
      return new UnicodeCategory(n.join(""));
    }, se = "\\P{", fe = '"\\\\P{"', pe = function(n) {
      return new UnicodeCategory(n.join(""), !0);
    }, he = function() {
      return new Token("null-character");
    }, de = 0, Ce = 0, Ae = 0, be = {
      line: 1,
      column: 1,
      seenCR: !1
    }, ve = 0, ke = [], ye = 0;
    if ("startRule" in vt) {
      if (!(vt.startRule in kt)) throw new Error("Can't start parsing from rule \"" + vt.startRule + '".');
      yt = kt[vt.startRule];
    }
    if (Token.offset = r, Token.text = l, bt = yt(), null !== bt && de === n.length) return bt;
    throw o(ke), Ce = Math.max(de, ve), new t(ke, Ce < n.length ? n.charAt(Ce) : null, Ce, u(Ce).line, u(Ce).column);
  }
  return n(t, Error), {
    SyntaxError: t,
    parse: l
  };
}(), index = 1, cgs = {};

exports = module.exports = parse, exports.Token = Token, exports.Alternate = Alternate, 
Alternate.prototype = Object.create(Token.prototype), Alternate.prototype.constructor = Alternate, 
exports.Match = Match, Match.prototype = Object.create(Token.prototype), Match.prototype.constructor = Match, 
exports.Group = Group, Group.prototype = Object.create(Token.prototype), Group.prototype.constructor = Group, 
exports.CaptureGroup = CaptureGroup, CaptureGroup.prototype = Object.create(Group.prototype), 
CaptureGroup.prototype.constructor = CaptureGroup, exports.Quantified = Quantified, 
Quantified.prototype = Object.create(Token.prototype), Quantified.prototype.constructor = Quantified, 
exports.Quantifier = Quantifier, Quantifier.prototype = Object.create(Token.prototype), 
Quantifier.prototype.constructor = Quantifier, exports.CharSet = CharSet, CharSet.prototype = Object.create(Token.prototype), 
CharSet.prototype.constructor = CharSet, exports.CharacterRange = CharacterRange, 
CharacterRange.prototype = Object.create(Token.prototype), CharacterRange.prototype.constructor = CharacterRange, 
exports.CharacterClass = CharacterClass, CharacterClass.prototype = Object.create(Token.prototype), 
CharacterClass.prototype.constructor = CharacterClass, exports.Literal = Literal, 
Literal.prototype = Object.create(Token.prototype), Literal.prototype.constructor = Literal, 
exports.Unicode = Unicode, Unicode.prototype = Object.create(Token.prototype), Unicode.prototype.constructor = Unicode, 
exports.UnicodeCategory = UnicodeCategory, UnicodeCategory.prototype = Object.create(Token.prototype), 
UnicodeCategory.constructor = Unicode, exports.Hex = Hex, Hex.prototype = Object.create(Token.prototype), 
Hex.prototype.constructor = Hex, exports.Octal = Octal, Octal.prototype = Object.create(Token.prototype), 
Octal.prototype.constructor = Octal, exports.BackReference = BackReference, BackReference.prototype = Object.create(Token.prototype), 
BackReference.prototype.constructor = BackReference, exports.ControlCharacter = ControlCharacter, 
ControlCharacter.prototype = Object.create(Token.prototype), ControlCharacter.prototype.constructor = ControlCharacter;
